File: notes/dynamic_programming/fibonacci_numbers.java
class fibonacci {
  static int fib(int n) {
    if (n <= 1)
      return n;
    return fib(n - 1) + fib(n - 2);
  }

  public static void main(String args[]) {
    int n = 9;
    System.out.println(fib(n));
  }
}

// Time Complexity: T(n) = T(n-1) + T(n-2) which is exponential
// Extra Space: O(n) if we consider the function call stack size


// Fibonacci Series using Dynamic Programming
class fibonacci {
  static int fib(int n) {
    /* Declare an array to store Fibonacci numbers. */
    int[] f = new int[n + 1];
    int i;

    /* 0th and 1st number of the series are 0 and 1*/
    f[0] = 0;
    f[1] = 1;

    for (i = 2; i <= n; i++) {
      /* Add the previous 2 numbers in the series
        and store it */
      f[i] = f[i - 1] + f[i - 2];
    }

    return f[n];
  }

  public static void main(String args[]) {
    int n = 9;
    System.out.println(fib(n));
  }
}

// Time Complexity: O(n) 
// Extra Space: O(n) if we consider the function call stack size

// Java program for Fibonacci Series using Space
// Optimized Method
class fibonacci {
  static int fib(int n) {
    int a = 0, b = 1, c;
    if (n == 0)
      return a;
    for (int i = 2; i <= n; i++) {
      c = a + b;
      a = b;
      b = c;
    }
    return b;
  }

  public static void main(String args[]) {
    int n = 9;
    System.out.println(fib(n));
  }
}

// Time complexity: O(n)
// Space complexity: O(1)

File: notes/dynamic_programming/fibonacci_numbers.js
// generate nth fibonnaci number via dynamic programming
// dynamic programming - break problem into subproblems and then
// recursively find solns to the subproblems
const fib = (n) => {
  // initialize fib map w/ seed values
  // use to store soln to subproblems (memoization)
  let map = [0, 1, 1];

  // closure will populate the map recursively
  const calc = (n) => {
    if (!map[n]) {
      // subproblem is f(n-1) + f(n-2)
      map[n] = calc(n - 1) + calc(n - 2);
    }
    // instead of recalculating same problems
    //just lookup previous soln in map
    return map[n];
  };
  return calc(n);
};

// fib seq: 1, 1, 2, 3, 5, 8, 13, 21...
// ex. problem tree w/o dynamic programming time complexity: O(2^n)
//    5
//   3  2
//  2 1 1 1
// 1 1

// no longer need to recalculate the two node
// time complexity: O(n)
// at most n recursive calls are made
console.log(fib(5)); // => 5
